core/trino-main/src/main/java/io/trino/operator/scalar/RowToJsonCast.java

```diff
@@ -73,11 +80,21 @@ protected ScalarFunctionImplementation specialize(FunctionBinding functionBindin
         checkCondition(canCastToJson(type), INVALID_CAST_ARGUMENT, "Cannot cast %s to JSON", type);
 
         List<Type> fieldTypes = type.getTypeParameters();
+        List<TypeSignatureParameter> typeSignatureParameters = type.getTypeSignature().getParameters();
+
+        List<String> fieldNames = new ArrayList<>(fieldTypes.size());
         List<JsonGeneratorWriter> fieldWriters = new ArrayList<>(fieldTypes.size());
         for (int i = 0; i < fieldTypes.size(); i++) {
-            fieldWriters.add(createJsonGeneratorWriter(fieldTypes.get(i)));
+            fieldNames.add(typeSignatureParameters.get(i).getNamedTypeSignature().getName().orElse("field" + i));
```

Comment:
If the field name is missing, let's use an empty string as a key (see the discussion on the main conversation thread)

Reply:
Also, move this into the `else` block of the `if` below. It's not needed for legacy mode